state-surgery/state/memory_db.go

```diff
@@ -0,0 +1,305 @@
+package state
+
+import (
+	"bytes"
+	"math/big"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+var _ vm.StateDB = (*MemoryStateDB)(nil)
+
+var emptyCodeHash = crypto.Keccak256(nil)
+
+type MemoryStateDB struct {
+	rw      sync.RWMutex
+	genesis *core.Genesis
+}
+
+func NewMemoryStateDB(genesis *core.Genesis) *MemoryStateDB {
+	if genesis == nil {
+		genesis = core.DeveloperGenesisBlock(15, 15_000_000, common.Address{})
+	}
+
+	return &MemoryStateDB{
+		genesis: genesis,
+		rw:      sync.RWMutex{},
+	}
+}
+
+//
+func (db *MemoryStateDB) Genesis() *core.Genesis {
+	return db.genesis
+}
+
+func (db *MemoryStateDB) GetAccount(addr common.Address) *core.GenesisAccount {
+	db.rw.RLock()
+	defer db.rw.RUnlock()
+
+	account, ok := db.genesis.Alloc[addr]
+	if !ok {
+		return nil
+	}
+	return &account
+}
+
+// StateDB interface
+
+func (db *MemoryStateDB) CreateAccount(addr common.Address) {
+	db.rw.Lock()
+	defer db.rw.Unlock()
+
+	db.genesis.Alloc[addr] = core.GenesisAccount{
+		Code:    []byte{},
+		Storage: make(map[common.Hash]common.Hash),
+		Balance: big.NewInt(0),
+		Nonce:   0,
+	}
+}
+
+func (db *MemoryStateDB) SubBalance(addr common.Address, amount *big.Int) {
+	db.rw.Lock()
+	defer db.rw.Unlock()
+
+	account, ok := db.genesis.Alloc[addr]
+	if !ok {
```

Comment:
Is it always the case that `AddBalance` and `SubBalance` are always called after `CreateAccount`? In the current implementation, calling these methods before `CreateAccount` will lead to a silent failure to set the appropriate balance.

Reply:
This matches the current behavior of the statedb in geth, we could panic in the case of it being called without the account existing to be more strict

Reply:
I think we should panic. It's a deviation from Geth but silent failures always scare me.

Reply:
I think we should panic. It's a deviation from Geth but silent failures always scare me.